eeps/0001-mkderivation-to-devshell.md

```diff
@@ -0,0 +1,149 @@
+{ name ? if originalArgs ? name then "${originalArgs.name}-dev-shell" else "nix-shell"
+, # a list of packages to add to the shell environment
+  packages ? [ ]
+, # propagate all the inputs from the given derivations
+  inputsFrom ? [ ]
+, buildInputs ? [ ]
+, nativeBuildInputs ? [ ]
+, propagatedBuildInputs ? [ ]
+, propagatedNativeBuildInputs ? [ ]
+, ...
+}@attrs:
+let
+  mergeInputs = name:
+    (originalArgs.${name} or [ ]) ++
+    (attrs.${name} or [ ]) ++
+    # 1. get all `{build,nativeBuild,...}Inputs` from the elements of `inputsFrom`
+    # 2. since that is a list of lists, `flatten` that into a regular list
+    # 3. filter out of the result everything that's in `inputsFrom` itself
+    # this leaves actual dependencies of the derivations in `inputsFrom`, but never the derivations themselves
+    (lib.subtractLists inputsFrom (lib.flatten (lib.catAttrs name inputsFrom)));
+
+  rest = builtins.removeAttrs attrs [
+    "name"
+    "packages"
+    "inputsFrom"
+    "buildInputs"
+    "nativeBuildInputs"
+    "propagatedBuildInputs"
+    "propagatedNativeBuildInputs"
+    "shellHook"
+  ];
+in originalArgs // {
+  inherit name;
+
+  buildInputs = mergeInputs "buildInputs";
+  nativeBuildInputs = packages ++ (mergeInputs "nativeBuildInputs");
+  propagatedBuildInputs = mergeInputs "propagatedBuildInputs";
+  propagatedNativeBuildInputs = mergeInputs "propagatedNativeBuildInputs";
+
+  # Avoid explicit checkout, and assume that shell will be used on source in repo
+  src = null;
+
+  shellHook = lib.concatStringsSep "\n" (lib.catAttrs "shellHook"
+    (lib.reverseList inputsFrom ++ [ attrs ]));
+
+  phases = [ "buildPhase" ];
+
+  buildPhase = ''
+    { echo "------------------------------------------------------------";
+      echo " WARNING: the existence of this path is not guaranteed.";
+      echo " It is an internal implementation detail for pkgs.mkShell.";
+      echo "------------------------------------------------------------";
+      echo;
+      # Record all build inputs as runtime dependencies
+      export;
+    } >> "$out"
+  '';
+
+  preferLocalBuild = true;
+} // rest
+```
+
+## Rust package example
+
+Instead of creating a separate dev shell, one can largely reuse the packaging
+logic from the nix expression.
+```nix
+# Before
+mkShell {
+  nativeBuildInputs = [
+    cargo
+    rustc
+    clippy
+  ];
+  ...
+}
+
+# After, re-using the packaging logic from myPackage expression
+myPackage.toDevShell {
+  nativeBuildInputs = [ clippy ];
+}
+
+# Stdenv can be passed to inspect compatibility
+myPackage.toDevShell (stdenv: {
+  nativeBuildInputs = lib.optionals stdenv.buildPlatform.isLinux [
+    gdb
+  ] ++ [
+    clippy
+  ];
+
+  buildInputs = lib.optionals stdenv.isDarwin [
+    CoreServices
+  ];
+})
+```
+
+## Unresolved issues or concerns
+
+- Adding a `.toDevShell` attr on every derivation does have a cost
+  - The expense of this should be carefully analyzed before fully adopting
+
+## Future work
+
+N/A
```

Comment:
we could also support targeting debug builds for things like cargo, cmake, meson, etc.